### PR TITLE
Drop use of GOBIN

### PR DESCRIPTION
**Pull request by isinyaaa**
Resolves: https://issues.redhat.com/browse/RHOAIENG-8808

## Description
Enable multi-arch builds by dropping GOBIN.

## Merge criteria:

- [ ] The commits and have meaningful messages; the author will squash them [after approval](https://github.com/opendatahub-io/opendatahub-community/blob/main/contributor-cheatsheet.md#:~:text=Usually%20this%20is%20done%20in%20last%20phase%20of%20a%20PR%20revision) or will ask to merge with squash.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -22,7 +22,14 @@ jobs:
...
         with:
-          go-version: '1.21'
+          go-version: "1.21"
+      - name: Install go dependencies
+        run: |
+          go install golang.org/dl/go1.21.9@latest
+          go install github.com/searKing/golang/tools/go-enum@v1.2.97
+          go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.31.0
+          go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.3.0
+          go install github.com/jmattheis/goverter/cmd/goverter@v1.4.1
```
Comment (tarilabs) [inline .github/workflows/build.yml:32]: I'm not convinced moving this from the makefile to the GHA is a good idea.

I like the avoidance of GOBIN, but having these "go install x" in the makefile avoided of the end-user manual action to ensure many prerequisites.
I can understand maybe a make target can no longer be identified easily, but in that case I'd move these go install x into a script file, which is invoked from make _and_ gha.

File: README.md [hunk 2]
```diff
@@ -10,11 +10,21 @@ Model registry provides a central repository for model developers to store and m
...
 - npm >= 10.2.0 - [Installing Node.js and npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm)
 - Java >= 11.0
 - python 3.9
 
+To install the go dependencies you can use
+
+```sh
+echo golang.org/dl/go1.21.9@latest github.com/searKing/golang/tools/go-enum@v1.2.97 google.golang.org/protobuf/cmd/protoc-gen-go@v1.31.0 google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.3.0 github.com/jmattheis/goverter/cmd/goverter@v1.4.1 | xargs -n1 go install
+```
+
```
Comment (tarilabs) [inline README.md:27]: see above related comment; I'd rather have this taken care somehow by the makefile invocation.

I see some project further automate this for GoLang by having toolchains listed in go.mod too, but didn't dig deeper.

File: scripts/build_deploy.sh [hunk 3]
```diff
@@ -1,13 +1,11 @@
...
+#!/usr/bin/env bash
 
 set -e
 
 # see Makefile for the IMG_ variables semantic
 IMG_REGISTRY=""
 IMG_ORG="${IMG_ORG:-kubeflow}"
 IMG_REPO="${IMG_REPO:-model-registry}"
-DOCKER_USER="${DOCKER_USER}"
-DOCKER_PWD="${DOCKER_PWD}"
```
Comment (tarilabs) [inline scripts/build_deploy.sh:10]: isn't this conventional to make it more evident these are the env variables required when someone is invoking this script, for instance from make/gha ?

File: Dockerfile [hunk 4]
```diff
@@ -9,32 +9,18 @@ COPY ["go.mod", "go.sum", "./"]
 RUN go mod download
 
 USER root
-# default NodeJS 14 is not enough for openapi-generator-cli, switch to Node JS currently supported
-RUN yum remove -y nodejs npm
-RUN yum module -y reset nodejs
-RUN yum module -y enable nodejs:18
-# install npm and java for openapi-generator-cli
-RUN yum install -y nodejs npm java-11
```
Comment (tarilabs) [inline Dockerfile:17]: these in the past proved "painful" to be done in docker merely for producing the end image, and I'm so glad to see them removed _from here_ 🥹

if we wanted easier DX we can always have these in a [DevContainer](https://containers.dev/) instead in the future, wdyt ?

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -22,7 +22,14 @@ jobs:
       - name: Setup Go
         uses: actions/setup-go@v5
         with:
-          go-version: '1.21'
+          go-version: "1.21"
+      - name: Install go dependencies
+        run: |
+          go install golang.org/dl/go1.21.9@latest
```
Comment (tarilabs) [inline .github/workflows/build.yml:28]:
```suggestion
```

I do not believe we still need this line, but kindly to confirm?

File: Makefile [hunk 6]
```diff
@@ -165,8 +148,8 @@ vendor:
 build: gen vet lint
 	${GO} build -buildvcs=false
 
-.PHONY: build/odh
-build/odh: vet
+.PHONY: build/docker
+build/docker: vet
 	${GO} build -buildvcs=false
```
Comment (tarilabs) [inline Makefile:153]: Can we do without `-buildvcs` now, or do we still need this option?

File: scripts/gen_openapi_server.sh [hunk 7]
```diff
@@ -1,28 +1,28 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 set -e
 
 echo "Generating the OpenAPI server"
 
-ROOT_FOLDER="${ROOT_FOLDER:-..}"
+PROJECT_ROOT=$(dirname "$(realpath "$0")")/..
```
Comment (tarilabs) [inline scripts/gen_openapi_server.sh:7]: what are your thought about using this as an absolute path, to make it consistent with https://github.com/kubeflow/model-registry/pull/256/files ?

File: scripts/gen_type_asserts.sh [hunk 8]
```diff
@@ -1,11 +1,11 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 set -e
 
-ROOT_FOLDER="${ROOT_FOLDER:-..}"
+PROJECT_ROOT=$(dirname "$(realpath "$0")")/..
```
Comment (tarilabs) [inline scripts/gen_type_asserts.sh:5]: ... same comment about absolute path here ...

File: scripts/install_protoc.sh [hunk 9]
```diff
@@ -1,22 +1,22 @@
...
+
 ARCH="x86_64"
-if [[ "$(uname -m)" == "arm"* ]]; then
-  ARCH="aarch_64"
+if [[ $(uname -m) =~ arm ]]; then
+    ARCH="aarch_64"
 fi
 
-mkdir -p ${SCRIPT_PARENT_DIR}/bin
+PROJECT_ROOT=$(dirname "$(realpath "$0")")/..
```
Comment (tarilabs) [inline scripts/install_protoc.sh:15]: ... the comment about absolute path is applying _mandatory_ here instead, for the reasons explained by @Al-Pragliola in https://github.com/kubeflow/model-registry/pull/256

File: scripts/build_deploy.sh (hunk 3 above)
Comment (isinyaaa) [reply scripts/build_deploy.sh:10]: I wasn't very sure tbh, we can keep them.

File: Makefile (hunk 6 above)
Comment (isinyaaa) [reply Makefile:153]: I managed to build it without this flag, let's see if that works on the CI as well.

File: scripts/gen_openapi_server.sh (hunk 7 above)
Comment (isinyaaa) [reply scripts/gen_openapi_server.sh:7]: good catch!

**Comment by isinyaaa** [comment]
> thank you so much @isinyaaa for this! 🚀
> 
> I left some comments, and in general I like how this simplifies the flow in general for the docker build too.
> 
> Can you kindly document as well, giving this a spin in producing an arm arch image, please?
> 
> To be specific, I'm _not_ asking to have the multi-arch done in this PR; I'm asking to give it a test so we're better confident with these modifications, we can go-compile for other archs _without_ the troubles we experienced in the past.

Simply changing the GOARCH variable works just fine :)

**Comment by tarilabs** [comment]
thank you so much @isinyaaa this make it make leaner imho and I'm glad to hear didn't give any trouble trying to supply now a different target arch :)

I leave to you to consider one additional option:

```diff
diff --git a/Makefile b/Makefile
index a28e850..5659b62 100644
--- a/Makefile
+++ b/Makefile
@@ -144,7 +144,7 @@ clean/deps:
        rm -Rf bin/*
 
 .PHONY: deps
-deps: bin/protoc bin/golangci-lint bin/openapi-generator-cli
+deps: go-deps bin/protoc bin/golangci-lint bin/openapi-generator-cli
 
 .PHONY: vendor
 vendor:
```

this would avoid having to launch the target manually as part of a build, or ensuring the versions we are intending for toolchains. wdyt? but again very minor.

I'm going to 
/lgtm
from my end 🚀 

I await others if they want try out this PR locally or if any remaining comments!

**Comment by isinyaaa** [comment]
> thank you so much @isinyaaa this make it make leaner imho and I'm glad to hear didn't give any trouble trying to supply now a different target arch :)
> 
> I leave to you to consider one additional option:
> 
> ```diff
> diff --git a/Makefile b/Makefile
> index a28e850..5659b62 100644
> --- a/Makefile
> +++ b/Makefile
> @@ -144,7 +144,7 @@ clean/deps:
>         rm -Rf bin/*
>  
>  .PHONY: deps
> -deps: bin/protoc bin/golangci-lint bin/openapi-generator-cli
> +deps: go-deps bin/protoc bin/golangci-lint bin/openapi-generator-cli
>  
>  .PHONY: vendor
>  vendor:
> ```
> 
> this would avoid having to launch the target manually as part of a build, or ensuring the versions we are intending for toolchains. wdyt? but again very minor.

I'm not completely against it as it would be "easier" but keep in mind that this adds ~10s to the plain `make build`, which is the default cmd. This "negates" the supposed benefits of the go compiler having super-fast builds, so maybe we could create another rule closer to `build/docker`, or simply change its name, to facilitate iteration. WDYT?

**Comment by tarilabs** [comment]
> I'm not completely against it as it would be "easier" but keep in mind that this adds ~10s to the plain `make build`, which is the default cmd. This "negates" the supposed benefits of the go compiler having super-fast builds, so maybe we could create another rule closer to `build/docker`, or simply change its name, to facilitate iteration. WDYT?

frankly 10s seems a bit exaggerated to me, I'm testing this one locally and if anything amounts to 2s, which in the scheme of the other generators time is negligible.

anyhow let's leave it off for now, since it's also documented prerequisite.
maybe time permitting for DX we could have some prechecks before issueing the `install`

**Comment by isinyaaa** [comment]
> > I'm not completely against it as it would be "easier" but keep in mind that this adds ~10s to the plain `make build`, which is the default cmd. This "negates" the supposed benefits of the go compiler having super-fast builds, so maybe we could create another rule closer to `build/docker`, or simply change its name, to facilitate iteration. WDYT?
> 
> frankly 10s seems a bit exaggerated to me, I'm testing this one locally and if anything amounts to 2s, which in the scheme of the other generators time is negligible.
> 
> anyhow let's leave it off for now, since it's also documented prerequisite. maybe time permitting for DX we could have some prechecks before issueing the `install`

It looks like this varies depending on the system cache. Re-running after a while will take longer than "immediate" re-runs.
We can change the default cmd behaviour later after this PR, either way it could be helpful if we had such a precheck mechanism for (potentially) system deps.
Going without the suggestion as it might be unwanted to change user-system deps.

**Comment by dhirajsb** [comment]
My biggest issue was the go binary in the bin directory. I thought that was an overkill. I was ok with having the other tool dependencies in `bin`. And, I've run into dependency compatibility issues with other projects, so having projects add their tool dependencies to my local go is a bit of an issue. 

Is the GOBIN really a hurdle for multi-platform builds? We can still download all the arch specific dependencies before building in the dockerfile for multiple platforms. 

I'm also concerned about the immutability of the Docker image build if it makes assumptions about what tools are available. That's the whole reason for a multi-stage build to avoid local env side affects. 

@lampajr wdyt?

**Comment by dhirajsb** [comment]
About local go and project go versions, we can also look into using go [toolchain](https://go.dev/doc/toolchain)

File: Dockerfile [hunk 10]
```diff
@@ -9,32 +9,18 @@ COPY ["go.mod", "go.sum", "./"]
...
 COPY pkg/ pkg/
-COPY patches/ patches/
-COPY templates/ templates/
-
-# Download tools
-RUN make deps
 
 # Build
-USER root
-RUN CGO_ENABLED=1 GOOS=linux GOARCH=amd64 make clean model-registry
```
[the remark this answers is not in the document]
Comment (isinyaaa) [reply Dockerfile:37]: Using this option also makes multi-arch builds much more complex, as you have to specify all the toolchain components by hand. I also couldn't find a reason to keep it, do you have any objection?

**Comment by isinyaaa** [comment]
> My biggest issue was the go binary in the bin directory. I thought that was an overkill. I was ok with having the other tool dependencies in `bin`. And, I've run into dependency compatibility issues with other projects, so having projects add their tool dependencies to my local go is a bit of an issue.
> 
> Is the GOBIN really a hurdle for multi-platform builds?

It is. While I'd recommend you try it yourself, feel free to take a look at this thread: https://github.com/golang/go/issues/57485.

> We can still download all the arch specific dependencies before building in the dockerfile for multiple platforms.

That would imply not using GOBIN inside the Dockerfile, actually. AFAIK you can still specify GOBIN when invoking commands locally, too.

> I'm also concerned about the immutability of the Docker image build if it makes assumptions about what tools are available. That's the whole reason for a multi-stage build to avoid local env side affects.

I'm not sure how to remediate that, unfortunately. I'll take a look at your suggestion for using the `toolchain` cmd.

File: Dockerfile (hunk 10 above)
[the remark this answers is not in the document]
Comment (isinyaaa) [reply Dockerfile]: We can discuss this separately, I'll push the change to a separate PR.

File: Dockerfile [hunk 11]
```diff
@@ -9,32 +9,18 @@ COPY ["go.mod", "go.sum", "./"]
...
 USER root
-# default NodeJS 14 is not enough for openapi-generator-cli, switch to Node JS currently supported
-RUN yum remove -y nodejs npm
-RUN yum module -y reset nodejs
-RUN yum module -y enable nodejs:18
-# install npm and java for openapi-generator-cli
-RUN yum install -y nodejs npm java-11
 
 # Copy the go source
-COPY ["Makefile", "main.go", ".openapi-generator-ignore", "openapitools.json", "./"]
```
[the remark this answers is not in the document]
Comment (isinyaaa) [reply Dockerfile]: Agreed, let's keep this. Reverted the change partially, wdyt?

**Review by tarilabs — COMMENTED** [review]
We are checking-in generated source files inside the project.
I believe this is a good practice.

I believe the Dockerfile as it stands now is re-generating source files inside the docker build, with the end results that we're shipping binaries which potentially we don't have the source code anymore (as we don't have track of the sources generated inside the docker build itself).

I preferred the original modification, per https://github.com/kubeflow/model-registry/pull/265#discussion_r1714689546

**Comment by tarilabs** [comment]
> My biggest issue was the go binary in the bin directory. I thought that was an overkill

 👉  https://github.com/kubeflow/model-registry/pull/288

**Comment by isinyaaa** [comment]
I can rebase whenever we agree on which path to follow, the current failure stems from the current go bin path missing on the dockerfiles.

**Comment by tarilabs** [comment]
> what is the resolution you guys found on this, as I see this is closed.

with more recent changes (ie as it stand currently on main default branch), we just discovered gobin looks currently not an impediment to crosscompilation, so we're now looking to keep it (hence not drop its current usage). More updates to follow

**Pull request closed without merging**